File: clients/rn/src/engine/between/WarmHandoff.ts
/**
 * Warm handoff protocol for device switching (~2s transfer).
 *
 * and.
 */

import type { BetweenClient } from './BetweenClient';

export interface ConversationTurn {
  role: string;
  content: string;
  timestamp: number;
}

export interface WarmHandoffContext {
  fromDid: string;
  toDid: string;
  activeRoomId: string;
  openConversationDids: string[];
  recentTurns: ConversationTurn[];
  vitalitySnapshot: Record<string, number>;
  currentTask: string | null;
  timestamp: number;
}

export type HandoffCallback = (context: WarmHandoffContext) => void;

export class WarmHandoffManager {
  private handoffCallback: HandoffCallback | null = null;
  private unsubscribe: (() => void) | null = null;

  constructor(
    private readonly between: BetweenClient,
    private readonly nodeId: string,
    private readonly familyId: string,
  ) {}

  /** Register a callback for incoming warm handoff. */
  onHandoffReceived(callback: HandoffCallback): void {
    this.handoffCallback = callback;
  }

  /** Start listening for incoming handoffs directed to this node. */
  startListening(): void {
    const subject = this.handoffSubject('*', this.nodeId);
    this.unsubscribe = this.between.subscribe(subject, (_subject, data) => {
      try {
        const context = JSON.parse(new TextDecoder().decode(data)) as WarmHandoffContext;
        this.handoffCallback?.(context);
      } catch {
        // Malformed handoff — skip
      }
    });
  }

  /** Stop listening. */
  stopListening(): void {
    this.unsubscribe?.();
    this.unsubscribe = null;
  }

  /** Initiate a warm handoff to a target node. */
  sendHandoff(context: WarmHandoffContext, targetNodeId: string): void {
    const data = new TextEncoder().encode(JSON.stringify(context));
    this.between.publish(this.handoffSubject(this.nodeId, targetNodeId), data);
  }

  private handoffSubject(src: string, dst: string): string {
    return `between.household.${this.familyId}.${src}.${dst}.soul.handoff`;
  }
}

File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/between/WarmHandoff.kt
package org.wyrdsekai.app.engine.between

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Warm handoff protocol for device switching (~2s transfer).
 *
 * When the user switches devices, the active bud packages its current context
 * and sends it via a directed Between message. The receiving bud restores
 * context and continues seamlessly.
 *
 * Subject: "between.household.{familyId}.{srcNodeId}.{dstNodeId}.soul.handoff"
 *
 * and.
 */
@Serializable
data class WarmHandoffContext(
    val fromDid: String,
    val toDid: String,
    val activeRoomId: String,
    val openConversationDids: List<String> = emptyList(),
    val recentTurns: List<ConversationTurn> = emptyList(),
    val vitalitySnapshot: Map<String, Float> = emptyMap(),
    val currentTask: String? = null,
    val timestamp: Long,
)

@Serializable
data class ConversationTurn(
    val role: String,
    val content: String,
    val timestamp: Long,
)

/**
 * Manages warm handoff initiation and reception.
 */
class WarmHandoffManager(
    private val between: BetweenClient,
    private val nodeId: String,
    private val familyId: String,
) {
    private val json = Json { ignoreUnknownKeys = true }
    private var handoffCallback: ((WarmHandoffContext) -> Unit)? = null
    private var unsubscribe: (() -> Unit)? = null

    /** Register a callback for incoming warm handoff. */
    fun onHandoffReceived(callback: (WarmHandoffContext) -> Unit) {
        handoffCallback = callback
    }

    /** Start listening for incoming handoffs directed to this node. */
    fun startListening() {
        val subject = handoffSubject("*", nodeId)
        unsubscribe = between.subscribe(subject) { _, data ->
            try {
                val context = json.decodeFromString<WarmHandoffContext>(data.decodeToString())
                handoffCallback?.invoke(context)
            } catch (_: Exception) {
                // Malformed handoff — skip
            }
        }
    }

    /** Stop listening. */
    fun stopListening() {
        unsubscribe?.invoke()
        unsubscribe = null
    }

    /** Initiate a warm handoff to a target node. */
    fun sendHandoff(context: WarmHandoffContext, targetNodeId: String) {
        val data = json.encodeToString(context).encodeToByteArray()
        between.publish(handoffSubject(nodeId, targetNodeId), data)
    }

    private fun handoffSubject(src: String, dst: String): String =
        "between.household.$familyId.$src.$dst.soul.handoff"
}
